File: Stacks_Queues_PQ_Heaps/Stacks/Largest_Rectangle_in_Histogram.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<pair<int, int>> nsel(vector<int> &nums) {
        vector<pair<int, int>> ans;
        stack<pair<int, int>> s;
        for(int i=0; i<nums.size(); i++) {
            while(!s.empty() && s.top().first >= nums[i]) s.pop();
            if(s.empty()) ans.push_back(make_pair(-1, -1));
            else ans.push_back(s.top());
            s.push(make_pair(nums[i], i));
        }
        return ans;
    }
    int largestRectangleArea(vector<int>& heights) {
        vector<int> rHeights;
        for(int i=heights.size()-1; i>=0; i--)  rHeights.push_back(heights[i]);
        vector<pair<int, int>> l = nsel(heights), r=nsel(rHeights);
        reverse(r.begin(), r.end());
        int ans = 0;
        for(int i=0; i<heights.size(); i++) {
            int val = heights[i]*(heights.size()-r[i].second-l[i].second-2);
            ans = max(ans, val);
        }
        return ans;
    }
};

File: Stacks_Queues_PQ_Heaps/Stacks/Largest_Rectangle_in_Histogram.py
from collections import deque
from typing import List

def nsel(nums):
    s = deque()
    ans = []
    for i, n in enumerate(nums):
        while s and s[-1][0]>=n:
            s.pop()
        if s:
            ans.append(s[-1])
        else:
            ans.append((-1, -1))
        s.append((n, i))
    return ans

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        l = nsel(heights)
        r = nsel(heights[::-1])
        r.reverse()
        ans = 0
        for i, h in enumerate(heights):
            val = h*(len(heights)-r[i][1]-l[i][1]-2)
            ans = max(ans, val)
        return ans